Getting Started guide for Rails 8 (#53846)
rails/rails: guides/source/getting_started.md

After reading this guide, you will know:

* How to install Rails, create a new Rails application, and connect your application to a database.
* The general layout of a Rails application.
* The basic principles of MVC (Model, View, Controller) and RESTful design.
* How to quickly generate the starting pieces of a Rails application.

> * How to deploy your app to production using Kamal.

Reviewer: Not sure "How to deploy your app to production" can be part of Getting Started guide.

The Rails philosophy includes two major guiding principles:

* **Don't Repeat Yourself:** DRY is a principle of software development which
  states that "Every piece of knowledge must have a single, unambiguous, authoritative
  representation within a system". By not writing the same information over and over
  again, our code is more maintainable, more extensible, and less buggy.
* **Convention Over Configuration:** Rails has opinions about the best way to do many
  things in a web application, and defaults to this set of conventions, rather than
  require that you specify minutiae through endless configuration files.

Creating a New Rails Project
----------------------------

TIP: You can create new Rails apps with a preconfigured Dev Container development environment. This
is the fastest way to get started with Rails.

> For instructions see [Getting Started with Dev Containers](getting_started_with_devcontainer.html)

Reviewer 2: given Rails come with `devcontainer` setup, should we not make it the default path to avoid OS-specific issues? 🤷‍♂️ 
(or at least keep the mention)

Author: I think we'll move this to the Install Ruby on Rails guide.

Reviewer 3: Makes sense in the Install guide.

Author: Added in 8f4918d4f0

In your code editor, open the migration Rails created for us so we can see what the migration does. This is located in `db/migrate/<timestamp>_create_products.rb`:

```ruby
class CreateProducts < ActiveRecord::Migration[8.0]
  def change
    create_table :products do |t|
      t.string :name

      t.timestamps
    end
  end
end
```

This migration is telling Rails to create a new database table named `products`.

NOTE: In contrast to the model above, Rails makes the database table names _plural_, because the database holds all of the instances of each model (i.e., You are creating a database of _products_).

The `create_table` block then defines which columns and types should be defined in this database table.

`t.string :name` tells Rails to create a column in the `products` table called `name` and set the type as `string`.

`t.timestamps` is a shortcut for defining two columns on your models: `created_at:datetime` and `updated_at:datetime`. You'll see these columns on most Active Record models in Rails and they are automatically set by Active Record when creating or updating records.

### Running Migrations

> Now that you have defined what change to make to the database, use the following command to run the migrations:

Reviewer 4: Now that you have defined what changes to make to the database

- `example.org` is the **host**
- `/products` is the **path**
- `?sale=true&sort=asc` are the **query parameters**

### HTTP Methods and Their Purpose

HTTP requests use methods to tell a server what action it should perform for a given URL. Here are the most common methods:

- A `GET` request tells the server to retrieve the data for a given URL (e.g., loading a page or fetching a record).
- A `POST` request will submit data to the URL for processing (usually creating a new record).
- A `PUT` or `PATCH` request submits data to a URL to update an existing record.
- A `DELETE` request to a URL tells the server to delete a record.

### Rails Routes

A `route` in Rails refers to a line of code that pairs an HTTP Method and a URL path. The route also tells Rails which `controller` and `action` should respond to a request.

To define a route in Rails, let's go back to your code editor and add the following route to `config/routes.rb`

```ruby
Rails.application.routes.draw do
  get "/products", to: "products#index"
end
```

This route tells Rails to look for GET requests to the `/products` path. In this example, we specified `"products#index"` for where to route the request.

> When Rails sees a request that matches, it will send the request to the `ProductsController` and the `index` action inside of that controller. This is how Rails will the request and return a response to the browser.

Reviewer 4: When Rails sees a request that matches, it will send the request to the `ProductsController` and the `index` action inside of that controller. This is how Rails will map the request and returns a response to the browser.

    <%= form.label :description, style: "display: block" %>
    <%= form.rich_text_area :description %>
  </div>

  <div>
    <%= form.submit %>
  </div>
<% end %>
```

Our controller also needs to permit this new parameter when the form is submitted, so we'll update the permitted params to include description in `app/controllers/products_controller.rb`

```ruby#3
    # Only allow a list of trusted parameters through.
    def product_params
      params.expect(product: [ :name, :description ])
    end
```

We also need to update the show view to display the description in `app/views/products/show.html.erb`:

```erb#3
<% cache @product do %>
  <h1><%= @product.name %></h1>
  <%= @product.description %>
<% end %>
```

The cache key generated by Rails also changes when the view is modified. This makes sure the cache stays in sync with the latest version of the view template.

Create a new product and add a description with bold and italic text. You'll see that the show page displays the formatted text and editing the product retains this rich text in the text area.

Check out the [Action Text Overview](action_text_overview.html) to learn more.

File Uploads with Active Storage

> -------------------------------

Reviewer 4: File Uploads with Active Storage
--------------------------------

> -------------------------------

Reviewer 4: File Uploads with Active Storage
--------------------------------

> -------------------------------

Reviewer 4: File Uploads with Active Storage
#### --------------------------------

> -------------------------------

Reviewer 4: missing a hyphen here

Introduction
------------

Welcome to Ruby on Rails! In this guide, we'll walk through the core concepts of building web applications with Rails. You don't need any experience with Rails to follow along with this guide.

Rails is a web framework built for the Ruby programming language. Rails takes advantage of many features of Ruby so we **strongly** recommend learning the basics of Ruby so that you understand some of the basic terms and vocabulary you will see in this tutorial.

- [Official Ruby Programming Language website](https://www.ruby-lang.org/en/documentation/)
- [List of Free Programming Books](https://github.com/EbookFoundation/free-programming-books/blob/master/books/free-programming-books-langs.md#ruby)

Rails Philosophy
----------------

Rails is a web application development framework written in the Ruby programming language. It is designed to make programming web applications easier by making assumptions about what every developer needs to get started. It allows you to write less code while accomplishing more than many other languages and frameworks. Experienced Rails developers also report that it makes web application development more fun.

> Rails is opinionated software. It makes the assumption that there is a "best" way to do things, and it's designed to encourage that way - and in some cases to discourage alternatives. If you learn "The Rails Way" you'll probably discover a tremendous increase in productivity. If you persist in bringing old habits from other languages to your Rails development, and trying to use patterns you learned elsewhere, you may have a less happy experience.

Reviewer 5: ```suggestion
Rails is an opinionated software. It makes the assumption that there is a "best" way to do things, and it's designed to encourage that way - and in some cases to discourage alternatives. If you learn "The Rails Way" you'll probably discover a tremendous increase in your productivity. If you persist in bringing old habits from other languages to your Rails development, and try to use patterns you learned elsewhere, you may have a less happy experience.
```

Author: Since this is what's in the current guide, I'm going to leave it as-is.

|test/|Unit tests, fixtures, and other test apparatus. These are covered in [Testing Rails Applications](testing.html).|
|tmp/|Temporary files (like cache and pid files).|
|vendor/|A place for all third-party code. In a typical Rails application this includes vendored gems.|
|.dockerignore|This file tells Docker which files it should not copy into the container.|
|.gitattributes|This file defines metadata for specific paths in a Git repository. This metadata can be used by Git and other tools to enhance their behavior. See the [gitattributes documentation](https://git-scm.com/docs/gitattributes) for more information.|
|.git/|Contains Git repository files.|
|.github/|Contains GitHub specific files.|
|.gitignore|This file tells Git which files (or patterns) it should ignore. See [GitHub - Ignoring files](https://help.github.com/articles/ignoring-files) for more information about ignoring files.|
|.kamal/|Contains Kamal secrets and deployment hooks.|
|.rubocop.yml|This file contains the configuration for RuboCop.|
|.ruby-version|This file contains the default Ruby version.|

### Model-View-Controller Basics

> Rails code is organized using the Model-View-Controller (MVC) architecture. With MVC, we have three main concepts where the majority of our code lives:

Reviewer 6: We could deep link to that pattern: 

https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller 

or 

https://www.w3schools.in/mvc-architecture

Reviewer 3: In general we try not to link out as much as possible (not always feasible), but we are creating a diagram for this section to help visualize MVC architecture, which should help folks understand it better (and keep them on the page). It should go live next week with the tutorial.

Reviewer 6: Nice! That makes sense.

Now that we've got a basic understanding of MVC, let's see how it's used in Rails.

Hello, Rails!
-------------

Let's start easy and boot up our Rails server for the first time.

In your terminal, run the following command in the `store` directory:

```bash
$ bin/rails server
```

This will start up a web server called Puma that will serve static files and your Rails application:

```bash
=> Booting Puma
=> Rails 8.0.0 application starting in development
=> Run `bin/rails server --help` for more startup options
Puma starting in single mode...
* Puma version: 6.4.3 (ruby 3.3.5-p100) ("The Eagle of Durango")
*  Min threads: 3
*  Max threads: 3
*  Environment: development
*          PID: 12345
* Listening on http://127.0.0.1:3000
* Listening on http://[::1]:3000
Use Ctrl-C to stop
```

To see your Rails application, open http://localhost:3000 in your browser. You will see the default Rails welcome page:

![Rails welcome page](images/getting_started/rails_welcome.png)

It works!

This page is the *smoke test* for a new Rails application, ensuring that everything is working behind the scenes to serve a page.

To stop the Rails server anytime, press `Ctrl-C` in your terminal.

> ### Automatic Reloading in Development

Reviewer 7: I got a bit excited and thought the exact same thing!

> ### Automatic Reloading in Development

Author: I've renamed this to `Autoloading in Development` to better clarify that.

Developer happiness is a cornerstone philosophy of Rails and one way of achieving this is with automatic code reloading in development.

Once you start the Rails server, new files or changes to existing files are detected and automatically loaded or reloaded as necessary. This allows you to focus on building without having to restart your Rails server after every change.

> You may also notice that Rails applications rarely use `require` statements. Rails uses naming conventions to require files automatically so you can focus on writing your application code.

Reviewer 2: what if your background is completely different, maybe you don't know how Ruby works either, therefore, spotting the difference may not be obvious.
I would keep this out, or move to the app folder structure and mention that Rails strongly recommends sticking to naming conventions to avoid `require`.

Reviewer 3: Or: "You may also notice that Rails applications rarely use `require` statements like you may have seen in other programming languages." 
(I swear we had something like this before.)

Author: Added @AmandaPerino's line. I think we had that earlier too but it must have gotten lost somewhere.

In your code editor, open the migration Rails created for us so we can see what the migration does. This is located in `db/migrate/<timestamp>_create_products.rb`:

```ruby
class CreateProducts < ActiveRecord::Migration[8.0]
  def change
    create_table :products do |t|
      t.string :name

      t.timestamps
    end
  end
end
```

This migration is telling Rails to create a new database table named `products`.

NOTE: In contrast to the model above, Rails makes the database table names _plural_, because the database holds all of the instances of each model (i.e., You are creating a database of _products_).

The `create_table` block then defines which columns and types should be defined in this database table.

`t.string :name` tells Rails to create a column in the `products` table called `name` and set the type as `string`.

`t.timestamps` is a shortcut for defining two columns on your models: `created_at:datetime` and `updated_at:datetime`. You'll see these columns on most Active Record models in Rails and they are automatically set by Active Record when creating or updating records.

### Running Migrations

> Now that you have defined what change to make to the database, use the following command to run the migrations:

Reviewer 4: ```suggestion
Now that you have defined what changes to make to the database, use the following command to run the migrations:
```

```bash
$ bin/rails db:migrate
```

This command checks for any new migrations and applies them to your database. Its output looks like this:

```bash
== 20240426151900 CreateProducts: migrating ===================================
-- create_table(:products)
   -> 0.0030s
== 20240426151900 CreateProducts: migrated (0.0031s) ==========================
```

TIP: If you make a mistake, you can run `rails db:rollback` to undo the last migration.

Rails Console
-------------

Now that we have created our products table, we can interact with it in Rails. Let's try it out.

For this, we're going to use a feature of Rails called the *console*.  The console is a helpful, interactive tool for testing our code in our Rails application.

```bash
$ bin/rails console
```

You will be presented with a prompt like the following:

```irb
Loading development environment (Rails 8.1.0)
store(dev)>
```

Here we can type code that will be executed when we hit `Enter`. Let's try printing out the Rails version:

```irb
store(dev)> Rails.version
=> "8.0.0"
```

It works!

> Active Record Model Basics
> -------------------------

Reviewer 8: ```suggestion
Active Record Model Basics
--------------------------
```

  updated_at: "2024-12-02 20:30:02.997261000 +0000">]
```

Alternatively, we can assign attributes and call `save` when we're ready to validate and save changes to the database.

Let's change the name "Shoes" back to "T-Shirt".

```irb
store(dev)> product = Product.find(1)
store(dev)> product.name = "T-Shirt"
=> "T-Shirt"
store(dev)> product.save
  TRANSACTION (0.1ms)  BEGIN immediate TRANSACTION /*application='Store'*/
  Product Update (0.2ms)  UPDATE "products" SET "name" = 'T-Shirt', "updated_at" = '2024-11-09 22:39:09.693548' WHERE "products"."id" = 1 /*application='Store'*/
  TRANSACTION (0.0ms)  COMMIT TRANSACTION /*application='Store'*/
=> true
```

### Deleting Records

The `destroy` method can be used to delete a record from the database.

```irb
store(dev)> product.destroy
  TRANSACTION (0.1ms)  BEGIN immediate TRANSACTION /*application='Store'*/
  Product Destroy (0.4ms)  DELETE FROM "products" WHERE "products"."id" = 1 /*application='Store'*/
  TRANSACTION (0.1ms)  COMMIT TRANSACTION /*application='Store'*/
=> #<Product:0x0000000125813d48 id: 1, name: "T-Shirt", created_at: "2024-11-09 22:39:38.498730000 +0000", updated_at: "2024-11-09 22:39:38.498730000 +0000">
```

> This deleted the T-Shirt product from our database. We can confirm this with `Product.all` to see that that it only returns Pants.

Reviewer 4: ```suggestion
This deleted the T-Shirt product from our database. We can confirm this with `Product.all` to see that it only returns Pants.
```

Let's add a `presence` validation to the Product model to ensure that all products must have a `name`.

```ruby
class Product < ApplicationRecord
  validates :name, presence: true
end
```

Let's try to create a Product without a name in the Rails console.

```irb
store(dev)> product = Product.new
store(dev)> product.save
=> false
```

This time `save` returns `false` because the `name` attribute wasn't specified.

Rails automatically runs validations during create, update, and save operations to ensure valid input. To see a list of errors generated by validations, we can call `errors` on the instance.

```irb
store(dev)> product.errors
=> #<ActiveModel::Errors [#<ActiveModel::Error attribute=name, type=blank, options={}>]>
```

This returns an `ActiveModel::Errors` object that can tell us exactly which errors are present.

It also can generate friendly error messages for us that we can use in our user interface.

```irb
store(dev)> product.errors.full_messages
=> ["Name can't be blank"]
```

Now let's build a web interface for our Products.

We are done with the console for now, so you can exit out of it by running `exit`.

> A Request's Journey Through Rails
> ---------------------

Reviewer 8: ```suggestion
A Request's Journey Through Rails
---------------------------------
```

- `example.org` is the **host**
- `/products` is the **path**
- `?sale=true&sort=asc` are the **query parameters**

### HTTP Methods and Their Purpose

HTTP requests use methods to tell a server what action it should perform for a given URL. Here are the most common methods:

- A `GET` request tells the server to retrieve the data for a given URL (e.g., loading a page or fetching a record).
- A `POST` request will submit data to the URL for processing (usually creating a new record).
- A `PUT` or `PATCH` request submits data to a URL to update an existing record.
- A `DELETE` request to a URL tells the server to delete a record.

### Rails Routes

A `route` in Rails refers to a line of code that pairs an HTTP Method and a URL path. The route also tells Rails which `controller` and `action` should respond to a request.

To define a route in Rails, let's go back to your code editor and add the following route to `config/routes.rb`

```ruby
Rails.application.routes.draw do
  get "/products", to: "products#index"
end
```

This route tells Rails to look for GET requests to the `/products` path. In this example, we specified `"products#index"` for where to route the request.

> When Rails sees a request that matches, it will send the request to the `ProductsController` and the `index` action inside of that controller. This is how Rails will the request and return a response to the browser.

Reviewer 4: ```suggestion
When Rails sees a request that matches, it will send the request to the `ProductsController` and the `index` action inside of that controller. This is how Rails will map/process the request and return a response to the browser.
```

Let's refactor this to use these helpers:

```erb#6
<h1>Products</h1>

<div id="products">
  <% @products.each do |product| %>
    <div>
      <%= link_to product.name, product %>
    </div>
  <% end %>
</div>
```

### Creating Products

So far we've had to create products in the Rails console, but let's make this work in the browser.

We need to create two actions for create:

1. The new product form to collect product information
2. The create action in the controller to save the product and check for errors

Let's start with our controller actions.

```ruby
class ProductsController < ApplicationController
  def index
    @products = Product.all
  end

  def show
    @product = Product.find(params[:id])
  end

  def new
    @product = Product.new
  end
end
```

The `new` action instantiates a new `Product` which we will use for displaying the form fields.

We can update `app/views/products/index.html.erb` to link to the new action.

```erb#3
<h1>Products</h1>

<%= link_to "New product", new_product_path %>

<div id="products">
  <% @products.each do |product| %>
    <div>

>         <%= link_to product.name, product %>

Reviewer 9: Looks like 4 space indentation here instead of just 2 :)

Author: Good catch! We had a mix of tabs and spaces from earlier reviews, so this probably left over from converting those.

To handle this, we first need to implement the `create` action in our controller.

```ruby#14,15,16,17,18,19,20,21,22,23,24,25,26,27
class ProductsController < ApplicationController
  def index
    @products = Product.all
  end

  def show
    @product = Product.find(params[:id])
  end

  def new
    @product = Product.new
  end

  def create
    @product = Product.new(product_params)
    if @product.save
      redirect_to @product
    else
      render :new, status: :unprocessable_entity
    end
  end

  private

    def product_params
      params.expect(product: [ :name ])
    end
end
```

#### Strong Parameters

The `create` action handles the data submitted by the form, but it needs to be filtered for security. That's where the `product_params` method comes into play.

In `product_params`, we tell Rails to inspect the params and ensure there is a key named `:product` with an array of parameters as the value. The only permitted parameters for products is `:name` and Rails will ignore any other parameters. This protects our application from malicious users who might try to hack our application.

> #### Handling errors

Reviewer 8: ```suggestion
#### Handling Errors
```

Add a small `<nav>` section inside the `<body>` with a link to Home and a Log out button.

```erb
<!DOCTYPE html>
<html>
  <!-- ... -->
  <body>
    <nav>
      <%= link_to "Home", root_path %>
      <%= button_to "Log out", session_path, method: :delete if authenticated? %>
    </nav>

    <%= yield %>
  </body>
</html>
```

This will display a Log out button only if the user is authenticated. When clicked, it will send a DELETE request to the session path which will log the user out.

### Allowing Unauthenticated Access

However, our store's product index and show pages should be accessible to everyone. By default, the Rails authentication generator will restrict all pages to authenticated users only.

To allow guests to view products, we can allow unauthenticated access in our controller.

```ruby
class ProductsController < ApplicationController
  allow_unauthenticated_access only: %i[ index show ]
  # ...
end
```

Log out and visit the products index and show pages to see they're accessible without being authenticated.

> ### Showing Links For Authenticated Users Only

Reviewer 8: ```suggestion
### Showing Links for Authenticated Users Only
```

Import maps keep the setup clean and minimal, while still supporting modern JavaScript features.

What are these JavaScript files already in our import map? They are a frontend framework called Hotwire that Rails uses by default.

### Hotwire

Hotwire is a JavaScript framework designed to take full advantage of server-side generated HTML. It is comprised of 3 core components:

1. [**Turbo**](https://turbo.hotwired.dev/) handles navigation, form submissions, page components, and updates without writing any custom JavaScript.
2. [**Stimulus**](https://stimulus.hotwired.dev/) provides a framework for when you need custom JavaScript to add functionality to the page.
3. [**Native**](https://native.hotwired.dev/) allows you to make hybrid mobile apps by embedding your web app and progressively enhancing it with native mobile features.

We haven't written any JavaScript yet, but we have been using Hotwire on the frontend. For instance, the form you created to add and edit a product was powered by Turbo.

Learn more in the [Asset Pipeline](asset_pipeline.html) and [Working with JavaScript in Rails](working_with_javascript_in_rails.html) guides.

> Rich Text Fields with Action Text
> --------------------------------

Reviewer 8: ```suggestion
Rich Text Fields with Action Text
---------------------------------
```

  <%# ... %>

  <div>
    <%= form.label :description, style: "display: block" %>
    <%= form.rich_text_area :description %>
  </div>

  <div>
    <%= form.submit %>
  </div>
<% end %>
```

Our controller also needs to permit this new parameter when the form is submitted, so we'll update the permitted params to include description in `app/controllers/products_controller.rb`

```ruby#3
    # Only allow a list of trusted parameters through.
    def product_params
      params.expect(product: [ :name, :description ])
    end
```

We also need to update the show view to display the description in `app/views/products/show.html.erb`:

```erb#3
<% cache @product do %>
  <h1><%= @product.name %></h1>
  <%= @product.description %>
<% end %>
```

The cache key generated by Rails also changes when the view is modified. This makes sure the cache stays in sync with the latest version of the view template.

Create a new product and add a description with bold and italic text. You'll see that the show page displays the formatted text and editing the product retains this rich text in the text area.

Check out the [Action Text Overview](action_text_overview.html) to learn more.

> File Uploads with Active Storage
> -------------------------------

Reviewer 4: ```suggestion
File Uploads with Active Storage
--------------------------------
```

The Rails philosophy includes two major guiding principles:

- **Don't Repeat Yourself:** DRY is a principle of software development which states that "Every piece of knowledge must have a single, unambiguous, authoritative representation within a system". By not writing the same information over and over again, our code is more maintainable, more extensible, and less buggy.
- **Convention Over Configuration:** Rails has opinions about the best way to do many things in a web application, and defaults to this set of conventions, rather than require that you define them yourself through endless configuration files.

Creating a New Rails App
------------------------

We're going to build a project called `store` - a simple e-commerce app that demonstrates several of Rails' built-in features.

> TIP: Any commands prefaced with a dollar sign `$` should be run in the terminal.

Reviewer 10: <img width="677" alt="image" src="https://github.com/user-attachments/assets/f8f07eab-9306-4b28-940e-be7a49b2fe02">

Taking this copy, making it a TIP and placing it right after "Creating a New Rails App" is such a GREAT UX.

It's very easy for us to remember this but it makes the guide accessible to people at the "Hello World" level. Love it.

### Prerequisites

For this project, you will need:

* Ruby 3.2 or newer

> * Rails 8.0.0 or newer

Reviewer: ```suggestion
* Rails 8.1.0 or newer
```

* A code editor

Follow the [Install Ruby on Rails Guide](install_ruby_on_rails.html) if you need to install Ruby and/or Rails.

Let's verify the correct version of Rails is installed. To display the current version, open a terminal and run the following. You should see a version number printed out:

```bash
$ rails --version

> Rails 8.1.0

Author: ```suggestion
Rails 8.0.0
```

```

> The version shown should be Rails 8.0.0 or higher.

Reviewer: ```suggestion
The version shown should be Rails 8.1.0 or higher.
```

Reviewer 11: Keep 8.0.0 in all of those. We can change to 8.1.0 after it is released

|.kamal/|Contains Kamal secrets and deployment hooks.|
|.rubocop.yml|This file contains the configuration for RuboCop.|
|.ruby-version|This file contains the default Ruby version.|

### Model-View-Controller Basics

Rails code is organized using the Model-View-Controller (MVC) architecture. With MVC, we have three main concepts where the majority of our code lives:

* Model - Manages the data in your application. Typically, your database tables.
* View - Handles rendering responses in different formats like HTML, JSON, XML, etc.
* Controller - Handles user interactions and the logic for each request.

Now that we've got a basic understanding of MVC, let's see how it's used in Rails.

Hello, Rails!
-------------

Let's start easy and boot up our Rails server for the first time.

In your terminal, run the following command in the `store` directory:

```bash
$ bin/rails server
```

This will start up a web server called Puma that will serve static files and your Rails application:

```bash
=> Booting Puma

> => Rails 8.0.0 application starting in development

Reviewer: ```suggestion
=> Rails 8.1.0 application starting in development
```

See [Autoloading and Reloading Constants](autoloading_and_reloading_constants.html) for more details.

Creating a Database Model
-------------------------

Active Record is a feature of Rails that maps relational databases to Ruby code. It helps generate the structured query language (SQL) for interacting with the database like creating, updating, and deleting tables and records. Our application is using SQLite which is the default for Rails.

Let's start by adding a database table to our Rails application to add products to our simple e-commerce store.

```bash
$ bin/rails generate model Product name:string
```

This command tells Rails to generate a model named `Product` which has a `name` column and type of `string` in the database. Later on, you'll learn how to add other column types.

You'll see the following in your terminal:

```bash
      invoke  active_record
      create    db/migrate/20240426151900_create_products.rb
      create    app/models/product.rb
      invoke    test_unit
      create      test/models/product_test.rb
      create      test/fixtures/products.yml
```

This command does several things. It creates...

> 1. a migration in the `db/migrate` folder

Reviewer: ```suggestion
1. A migration in the `db/migrate` folder.
```

> 2. an Active Record model in `app/models/product.rb`

Reviewer: ```suggestion
2. An Active Record model in `app/models/product.rb`.
```

> 3. tests and test fixtures for this model

Reviewer: ```suggestion
3. Tests and test fixtures for this model.
```

`t.string :name` tells Rails to create a column in the `products` table called `name` and set the type as `string`.

`t.timestamps` is a shortcut for defining two columns on your models: `created_at:datetime` and `updated_at:datetime`. You'll see these columns on most Active Record models in Rails and they are automatically set by Active Record when creating or updating records.

### Running Migrations

Now that you have defined what changes to make to the database, use the following command to run the migrations:

```bash
$ bin/rails db:migrate
```

This command checks for any new migrations and applies them to your database. Its output looks like this:

```bash
== 20240426151900 CreateProducts: migrating ===================================
-- create_table(:products)
   -> 0.0030s
== 20240426151900 CreateProducts: migrated (0.0031s) ==========================
```

TIP: If you make a mistake, you can run `bin/rails db:rollback` to undo the last migration.

Rails Console
-------------

Now that we have created our products table, we can interact with it in Rails. Let's try it out.

> For this, we're going to use a feature of Rails called the *console*.  The console is a helpful, interactive tool for testing our code in our Rails application.

Reviewer: ```suggestion
For this, we're going to use a Rails feature called the *console*. The console is a helpful, interactive tool for testing our code in our Rails application.
```

```bash
$ bin/rails console
```

You will be presented with a prompt like the following:

```irb

> Loading development environment (Rails 8.1.0)

Reviewer: nit: version

Author: ```suggestion
Loading development environment (Rails 8.0.0)
```

store(dev)>
```

Here we can type code that will be executed when we hit `Enter`. Let's try printing out the Rails version:

```irb
store(dev)> Rails.version

> => "8.0.0"

Reviewer: ```suggestion
=> "8.1.0"
```

```

The `product` variable is an instance of `Product`. It has not been saved to the database, and so does not have an ID, created_at, or updated_at timestamps.

We can call `save` to write the record to the database.

```irb
store(dev)> product.save
  TRANSACTION (0.1ms)  BEGIN immediate TRANSACTION /*application='Store'*/
  Product Create (0.9ms)  INSERT INTO "products" ("name", "created_at", "updated_at") VALUES ('T-Shirt', '2024-11-09 16:35:01.117836', '2024-11-09 16:35:01.117836') RETURNING "id" /*application='Store'*/
  TRANSACTION (0.9ms)  COMMIT TRANSACTION /*application='Store'*/
=> true
```

When `save` is called, Rails takes the attributes in memory and generates an `INSERT` SQL query to insert this record into the database.

Rails also updates the object in memory with the database record `id` along with the `created_at` and `updated_at` timestamps. We can see that by printing out the `product` variable.

```irb
store(dev)> product
=> #<Product:0x00000001221f6260 id: 1, name: "T-Shirt", created_at: "2024-11-09 16:35:01.117836000 +0000", updated_at: "2024-11-09 16:35:01.117836000 +0000">
```

> Similar to `save`, we can use `create` to instantiate and save an Active Record model in a single call.

Reviewer: ```suggestion
Similar to `save`, we can use `create` to instantiate and save an Active Record object in a single call.
```

```erb#4
<h1><%= @product.name %></h1>

<%= link_to "Back", products_path %>
<%= link_to "Edit", edit_product_path(@product) %>
```

#### Before Actions

Since `edit` and `update` require an existing database record like `show` we can deduplicate this into a `before_action`.

A `before_action` allows you to extract shared code between actions and run it *before* the action. In the above controller code, `@product = Product.find(params[:id])` is defined in three different methods. Extracting this query to a before action called `set_product` cleans up our code for each action.

This is a good example of the DRY (Don't Repeat Yourself) philosophy in action.

```ruby#2,8-9,24-25,27-33,37-39
class ProductsController < ApplicationController
  before_action :set_product, only: %i[ show edit update ]

  def index
    @products = Product.all
  end

  def show
  end

  def new
    @product = Product.new
  end

  def create
    @product = Product.new(product_params)
    if @product.save
      redirect_to @product
    else
      render :new, status: :unprocessable_entity
    end
  end

  def edit
  end

  def update
    if @product.update(product_params)
      redirect_to @product
    else

>       render :new, status: :unprocessable_entity

Reviewer 12: ```suggestion
      render :edit, status: :unprocessable_entity
```

```erb#3
<h1>Edit product</h1>

<%= render "form", product: @product %>
<%= link_to "Cancel", @product %>
```

To learn more about view partials, check out the [Action View Guide](action_view_overview.html).

### Deleting Products

The last feature we need to implement is deleting products. We will add a `destroy` action to our `ProductsController` to handle `DELETE /products/:id` requests.

Adding `destroy` to `before_action :set_product` let's us set the `@product` instance variable in the same way we do for the other actions.

```ruby#2,35-38
class ProductsController < ApplicationController
  before_action :set_product, only: %i[ show edit update destroy ]

  def index
    @products = Product.all
  end

  def show
  end

  def new
    @product = Product.new
  end

  def create
    @product = Product.new(product_params)
    if @product.save
      redirect_to @product
    else
      render :new, status: :unprocessable_entity
    end
  end

  def edit
  end

  def update
    if @product.update(product_params)
      redirect_to @product
    else

>       render :new, status: :unprocessable_entity

Reviewer 12: ```suggestion
      render :edit, status: :unprocessable_entity
```

  <div>
    <label for="product_name">Name</label>
    <input type="text" name="product[name]" id="product_name">
  </div>

  <div>
    <input type="submit" name="commit" value="Create Product" data-disable-with="Create Product">
  </div>
</form>
```

The form builder has included a CSRF token for security, configured the form for UTF-8 support, set the input field names and even added a disabled state for the submit button.

Because we passed a new `Product` instance to the form builder, it automatically generated a form configured to send a `POST` request to `/products`, which is the default route for creating a new record.

To handle this, we first need to implement the `create` action in our controller.

```ruby#14-27
class ProductsController < ApplicationController
  def index
    @products = Product.all
  end

  def show
    @product = Product.find(params[:id])
  end

  def new
    @product = Product.new
  end

  def create
    @product = Product.new(product_params)
    if @product.save
      redirect_to @product
    else
      render :new, status: :unprocessable_entity
    end
  end

  private

> 

Reviewer 11: ```suggestion
```

We also set the HTTP status to 422 Unprocessable Entity to tell the browser this POST request failed and to handle it accordingly.

### Editing Products

The process of editing records is very similar to creating records. Instead of `new` and `create` actions, we will have `edit` and `update`.

Let's implement them in the controller with the following:

```ruby#23-34
class ProductsController < ApplicationController
  def index
    @products = Product.all
  end

  def show
    @product = Product.find(params[:id])
  end

  def new
    @product = Product.new
  end

  def create
    @product = Product.new(product_params)
    if @product.save
      redirect_to @product
    else
      render :new, status: :unprocessable_entity
    end
  end

  def edit
    @product = Product.find(params[:id])
  end

  def update
    @product = Product.find(params[:id])
    if @product.update(product_params)
      redirect_to @product
    else

>       render :new, status: :unprocessable_entity

Author: Good catch, should be `:edit`. 

Want to use the Suggestion feature to change it?

>       render :new, status: :unprocessable_entity

Reviewer 12: ```suggestion
      render :edit, status: :unprocessable_entity
```

    end
  end

  private

> 

Reviewer 11: ```suggestion
```

Reviewer 11: The generator, and our styleguide doesn't have an empty line after the `private`

<%= link_to "Edit", edit_product_path(@product) %>
```

#### Before Actions

Since `edit` and `update` require an existing database record like `show` we can deduplicate this into a `before_action`.

A `before_action` allows you to extract shared code between actions and run it *before* the action. In the above controller code, `@product = Product.find(params[:id])` is defined in three different methods. Extracting this query to a before action called `set_product` cleans up our code for each action.

This is a good example of the DRY (Don't Repeat Yourself) philosophy in action.

```ruby#2,8-9,24-25,27-33,37-39
class ProductsController < ApplicationController
  before_action :set_product, only: %i[ show edit update ]

  def index
    @products = Product.all
  end

  def show
  end

  def new
    @product = Product.new
  end

  def create
    @product = Product.new(product_params)
    if @product.save
      redirect_to @product
    else
      render :new, status: :unprocessable_entity
    end
  end

  def edit
  end

  def update
    if @product.update(product_params)
      redirect_to @product
    else
      render :new, status: :unprocessable_entity
    end
  end

  private

> 

Reviewer 11: ```suggestion
```

<h1>Edit product</h1>

<%= render "form", product: @product %>
<%= link_to "Cancel", @product %>
```

To learn more about view partials, check out the [Action View Guide](action_view_overview.html).

### Deleting Products

The last feature we need to implement is deleting products. We will add a `destroy` action to our `ProductsController` to handle `DELETE /products/:id` requests.

Adding `destroy` to `before_action :set_product` let's us set the `@product` instance variable in the same way we do for the other actions.

```ruby#2,35-38
class ProductsController < ApplicationController
  before_action :set_product, only: %i[ show edit update destroy ]

  def index
    @products = Product.all
  end

  def show
  end

  def new
    @product = Product.new
  end

  def create
    @product = Product.new(product_params)
    if @product.save
      redirect_to @product
    else
      render :new, status: :unprocessable_entity
    end
  end

  def edit
  end

  def update
    if @product.update(product_params)
      redirect_to @product
    else
      render :new, status: :unprocessable_entity
    end
  end

  def destroy
    @product.destroy
    redirect_to products_path
  end

  private

> 

Reviewer 11: ```suggestion
```

Optionally, you can include a link to this route in the navbar to add a Login link if not authenticated.

```erb
<%= link_to "Login", new_session_path unless authenticated? %>
```

You can also update the Edit and Destroy links on the show view to only display if authenticated.

```erb#1,4
<% if authenticated? %>
  <%= link_to "Edit", edit_product_path(@product) %>
  <%= button_to "Destroy", @product, method: :delete, data: { turbo_confirm: "Are you sure?" } %>
<% end %>
```

Caching Products
----------------

Sometimes caching specific parts of a page can improve performance. Rails simplifies this process with Solid Cache, a database-backed cache store that comes included by default.

Using the `cache` method, we can store HTML in the cache. Let's cache the header in `app/views/products/show.html.erb`.

```erb#1,3
<% cache @product do %>
  <h1><%= @product.name %></h1>
<% end %>
```

By passing `@product` into `cache`, Rails generates a unique cache key for the product. Active Record objects have a `cache_key` method that returns a String like `"products/1"`. The `cache` helper in the views combines this with the template digest to create a unique key for this HTML.

> To enable caching in development, run `rails dev:cache` in your terminal.

Reviewer 11: ```suggestion
To enable caching in development, run `bin/rails dev:cache` in your terminal.
```

I didn't check everything but we should keep the commands consistent

2. [**Stimulus**](https://stimulus.hotwired.dev/) provides a framework for when you need custom JavaScript to add functionality to the page.
3. [**Native**](https://native.hotwired.dev/) allows you to make hybrid mobile apps by embedding your web app and progressively enhancing it with native mobile features.

We haven't written any JavaScript yet, but we have been using Hotwire on the frontend. For instance, the form you created to add and edit a product was powered by Turbo.

Learn more in the [Asset Pipeline](asset_pipeline.html) and [Working with JavaScript in Rails](working_with_javascript_in_rails.html) guides.

Rich Text Fields with Action Text
---------------------------------

Many applications need rich text with embeds (i.e. multimedia elements) and Rails provides this functionality out of the box with Action Text.

To use Action Text, you'll first run the installer:

```bash
$ bin/rails action_text:install
$ bundle install
$ bin/rails db:migrate
```

Restart your Rails server to make sure all the new features are loaded.

Now, let's add a rich text description field to our product.

First, add the following to the `Product` model:

```ruby#2
class Product

>   has_rich_text :description

Reviewer 11: Where did we create the description column? Would this work without the column?

Author: This is using ActionText, so it doesn't need it's own column since it uses the ActionText tables.

#### Extracting Partials

We've already written a form for creating new products. Wouldn't it be nice if
we could reuse that for edit and update? We can, using a feature called
"partials" that allows you to reuse a view in multiple places.

We can move the form into a file called `app/views/products/_form.html.erb`. The
filename starts with an underscore to denote this is a partial.

We also want to replace any instance variables with a local variable, which we

> can define when we render the partial. We'll replace `@product` with `product`.

Reviewer 13: Might be nice to give a brief explanation, for example:

We also want to replace any instance variables with a local variable so the partial can be used with different views and display different data that is passed to it.

This can be defined when we render the partial. We'll replace `@product` with `product`.

Author: Thanks @elhalvers. We discussed this a while back and didn't have a great explanation at the time. I can run with your suggestion. 👍

  <body>
    <div class="notice"><%= notice %></div>
    <!-- ... -->
  </body>
</html>
```

To subscribe users to a specific product, we'll use a nested route so we know which product the subscriber belongs to. In `config/routes.rb` change `resources :products` to the following:

```ruby
  resources :products do
    resources :subscribers, only: [ :create ]
  end
```

On the product show page, we can check if there is inventory and display the amount in stock. Otherwise, we can display an out of stock message with the subscribe form to get notified when it is back in stock.

Create a new partial at `app/views/products/_inventory.html.erb` and add the following:

```erb
<% if product.inventory_count? %>
  <p><%= product.inventory_count %> in stock</p>
<% else %>
  <p>Out of stock</p>
  <p>Email me when available.</p>

  <%= form_with model: [product, Subscriber.new] do |form| %>
    <%= form.email_field :email, placeholder: "you@example.com", required: true %>
    <%= form.submit "Submit" %>
  <% end %>
<% end %>
```

Then update `app/views/products/show.html.erb` to render this partial after the `cache` block.

```erb
<%= render "inventory", product: @product %>
```

> ### In stock email notifications

Reviewer 8: ```suggestion
### In Stock Email Notifications
```

  has_rich_text :description
  has_many :subscribers, dependent: :destroy

  validates :name, presence: true
  validates :inventory_count, numericality: { greater_than_or_equal_to: 0 }

  after_update_commit :notify_subscribers, if: :back_in_stock?

  def back_in_stock?
    inventory_count_previously_was.zero? && inventory_count > 0
  end

  def notify_subscribers
    subscribers.each do |subscriber|
      ProductMailer.with(product: self, subscriber: subscriber).in_stock.deliver_later
    end
  end
end
```

`after_update_commit` is an Active Record callback that is fired after changes are saved to the database. `if: :back_in_stock?` tells the callback to run only if the `back_in_stock?` method returns true.

Active Record keeps track of changes to attributes so `back_in_stock?` checks the previous value of `inventory_count` using `inventory_count_previously_was`. Then we can compare that against the current inventory count to determine if the product is back in stock.

`notify_subscribers` uses the Active Record association to query the `subscribers` table for all subscribers for this specific product and then queues up the `in_stock` email to be sent to each of them.

> ### Extracting A Concern

Reviewer 8: ```suggestion
### Extracting a Concern
```

.

Finished in 0.343842s, 2.9083 runs/s, 5.8166 assertions/s.
1 runs, 2 assertions, 0 failures, 0 errors, 0 skips
```

Our test passes!

Rails also generated an example test for `ProductMailer` at `test/mailers/product_mailer_test.rb`. Let's update it to make it also pass.

```ruby
require "test_helper"

class ProductMailerTest < ActionMailer::TestCase
  test "in_stock" do
    mail = ProductMailer.with(product: products(:tshirt), subscriber: subscribers(:david)).in_stock
    assert_equal "In stock", mail.subject
    assert_equal [ "david@example.org" ], mail.to
    assert_equal [ "from@example.com" ], mail.from
    assert_match "Good news!", mail.body.encoded
  end
end
```

Let's run the entire test suite now and ensure all the tests pass.

```bash
$ bin/rails test
Running 2 tests in a single process (parallelization threshold is 50)
Run options: --seed 16302

# Running:

..

Finished in 0.665856s, 3.0037 runs/s, 10.5128 assertions/s.
2 runs, 7 assertions, 0 failures, 0 errors, 0 skips
```

You can use this as a starting place to continue building out a test suite with full coverage of the application features.

Learn more about [Testing Rails Applications](testing.html)

> Consistently formatted code with Rubocop

Reviewer 8: ```suggestion
Consistently Formatted Code with Rubocop
```

```

```ruby
store(prod)> User.create!(email_address: "you@example.org", password: "s3cr3t", password_confirmation: "s3cr3t")
```

Now you can log in to production with this email and password and manage products.

### Background Jobs using Solid Queue

Background jobs allow you to run tasks asynchronously behind-the-scenes in a separate process, preventing them from interrupting the user experience. Imagine sending in stock emails to 10,000 recipients. It could take a while, so we can offload that task to a background job to keep the Rails app responsive.

In development, Rails uses the `:async` queue adapter to process background jobs with ActiveJob. Async stores pending jobs in memory but it will lose pending jobs on restart. This is great for development, but not production.

To make background jobs more robust, Rails uses `solid_queue` for production environments. Solid Queue stores jobs in the database and executes them in a separate process.

Solid Queue is enabled for our production Kamal deployment using the `SOLID_QUEUE_IN_PUMA: true` environment variable to `config/deploy.yml`. This tells our web server, Puma, to start and stop the Solid Queue process automatically.

> When emails are sent with Action Mailer's `deliver_later`, these emails will be sent to ActiveJob for sending in the background so they don't delay the HTTP request. With Solid Queue in production, emails will be sent in the background, automatically retried if they fail to send, and jobs are kept safe in the database during restarts.

Reviewer 11: ```suggestion
When emails are sent with Action Mailer's `deliver_later`, these emails will be sent to Active Job for sending in the background so they don't delay the HTTP request. With Solid Queue in production, emails will be sent in the background, automatically retried if they fail to send, and jobs are kept safe in the database during restarts.
```
